Hot Restart Encryption - OS part

vojtechtoman opened the pull request:
Introduced Encryption at Rest and Secure Store configuration
classes/properties and extended the core and Spring XML config
schemas accordingly.

Refactored out the core of `CipherHelper` (EE) into
`BasicSymmetricCipherBuilder` to make encrypt/decrypt functionality
available in the hotrestart module.

EE counterpart: https://github.com/hazelcast/hazelcast-enterprise/pull/3148 (original one is now obsolete and will be closed: https://github.com/hazelcast/hazelcast-enterprise/pull/3083)

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/internal/util/BasicSymmetricCipherBuilder.java:
```diff
@@ -0,0 +1,148 @@
+import static com.hazelcast.internal.memory.impl.EndiannessUtil.BYTE_ARRAY_ACCESS;
+import static com.hazelcast.util.Preconditions.checkNotNull;
+
+public class BasicSymmetricCipherBuilder {
```
Can we move this builder and the test into EE?

vojtechtoman replied:
Yes, we could, and I was actually thinking about that, too. It would have to reside in the hot-restart module but if that is fine, I can do that.

vojtechtoman replied:
Moved to the hot-restart module.

mmedenjak commented:
@kwart can you review the encryption parts, also in EE?

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/config/JavaKeyStoreSecureStoreConfig.java:
```diff
@@ -0,0 +1,225 @@
+     */
+    public static final String DEFAULT_KEYSTORE_TYPE = "JCEKS";
+
+    private File path;
```
Good catch, I will try to address this case.

vojtechtoman replied:
`path` is now passed in the constructor, `password` can be null (some keystore types support that)

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/config/ConfigXmlGenerator.java:
```diff
@@ -1304,16 +1303,31 @@ private static void socketInterceptorConfigXmlGenerator(XmlGenerator gen, Socket
+    }
+
+    private void symmetricEncInterceptorConfigXmlBodyGenerator(XmlGenerator gen, SymmetricEncryptionConfig sec) {
+        if (sec == null) {
```
+1, will do

vojtechtoman replied:
fixed

kwart commented on hazelcast/src/main/java/com/hazelcast/config/JavaKeyStoreSecureStoreConfig.java:
```diff
@@ -0,0 +1,237 @@
+    /**
+     * The default Java KeyStore type (JCEKS).
+     */
+    public static final String DEFAULT_KEYSTORE_TYPE = "JCEKS";
```
We should use the `PKCS12` instead. It's the default type from Java 9 onwards.
Even the `keytool` in newer Java versions suggest switching to `PKCS12` when you are creating `JKS` or `JCEKS` keystores:

```
Warning:
The JCEKS keystore uses a proprietary format. It is recommended to migrate to PKCS12 which is an industry standard format using "keytool -importkeystore -srckeystore vault.keystore -destkeystore vault.keystore -deststoretype pkcs12".
```

I also remember there are some problems with transferring JCEKS from one Java to another (IBM vs Oracle java).

The default keystore type up to Java 8 was `JKS`, but this type doesn't support secret keys. It's able to only handle certificates and keypairs.

vojtechtoman replied:
Thanks for the input. I will change the default to `PKCS12`.

vojtechtoman replied:
Default changed to `PKCS12`

kwart commented on hazelcast/src/main/java/com/hazelcast/config/JavaKeyStoreSecureStoreConfig.java:
```diff
@@ -0,0 +1,237 @@
+     * @throws IllegalArgumentException if password is {code null}
+     */
+    public JavaKeyStoreSecureStoreConfig setPassword(String password) {
+        checkNotNull(password, "Java Key Store password cannot be null!");
```
There are keystore types which don't require a password or gets it from another channel (e.g. `WINDOWS-MY` type).

vojtechtoman replied:
Indeed, good catch!

vojtechtoman replied:
null passwords are allowed now

kwart commented on hazelcast/src/main/java/com/hazelcast/config/JavaKeyStoreSecureStoreConfig.java:
```diff
@@ -0,0 +1,237 @@
+     */
+    public JavaKeyStoreSecureStoreConfig(File path, String password) {
+        checkNotNull(path, "Java Key Store path cannot be null!");
+        checkNotNull(password, "Java Key Store password cannot be null!");
```
Password may be `null`.

vojtechtoman replied:
Yes, I will relax the check.

vojtechtoman replied:
fixed

kwart commented on hazelcast/src/main/resources/hazelcast-config-4.0.xsd:
```diff
@@ -3661,6 +3661,222 @@
+                </xs:documentation>
+            </xs:annotation>
+        </xs:attribute>
+        <xs:attribute name="password" type="xs:string" use="required">
```
Usually, entry passwords are optional. When it's not provided, the password from the keystore itself is used.

vojtechtoman replied:
+1

vojtechtoman replied:
Entries are gone now.

kwart commented on hazelcast/src/main/resources/hazelcast-config-4.0.xsd:
```diff
@@ -3661,6 +3661,222 @@
+                    </xs:documentation>
+                </xs:annotation>
+            </xs:element>
+            <xs:element name="password" type="xs:string" minOccurs="1" maxOccurs="1">
```
Optional. Some KS types don't require passwords.

vojtechtoman replied:
Yes, will relax that.

vojtechtoman replied:
fixed

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/config/AbstractBasicSymmetricEncryptionConfig.java:
```diff
@@ -0,0 +1,68 @@
+        return enabled;
+    }
+
+    public T setEnabled(boolean enabled) {
```
fixed

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/config/VaultSecureStoreConfig.java:
```diff
@@ -0,0 +1,304 @@
+        return address;
+    }
+
+    public VaultSecureStoreConfig setAddress(String address) {
```
fixed

vojtechtoman commented on hazelcast/src/main/java/com/hazelcast/config/HotRestartPersistenceConfig.java:
```diff
@@ -230,6 +232,16 @@ public HotRestartPersistenceConfig setAutoRemoveStaleData(boolean autoRemoveStal
         return this;
     }
 
+    public HotRestartPersistenceConfig setEncryptionAtRestConfig(EncryptionAtRestConfig encryptionAtRestConfig) {
```
fixed

vojtechtoman commented:
I have pushed a new commit. The most notable changes (besides addressing the review comments) are:
- Introduction of an internal Secure Store SPI (moved out from the EE hot-restart module).
- Removal of the "entry" concept from both the KeyStore/Vault configurations.
- `JavaKeyStoreSecureStoreConfig`: added an option to specify the 'current key' entry alias.
- `VaultSecureStoreConfig`: removed the secrets engine version (detected automatically now) and the support for namespaces (equivalent functionality is available via the secret path). Made the address, access token, and secret path required (specified in the constructor).

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/spi/impl/securestore/SecureStore.java:
```diff
@@ -0,0 +1,67 @@
+ * A Secure Store abstraction for interacting with systems that manage symmetric
+ * encryption keys.
+ */
+public interface SecureStore extends Disposable {
```
Minor: how come these classes are in OS?

vojtechtoman replied:
Well, it's not a minor remark, it's actually a valid question. The goal was to make the secure store abstraction independent of hot-restart. I could have placed the APIs in a different package in the hot-restart submodule but I thought this was cleaner. At any rate, the top-level EE module (where the actual secure store implementations live) needs access to the APIs from some lower-level layer (= either the hot-restart submodule or OS).

We do something similar in OS with `HotRestartService` (which is public in OS), so I thought a similar approach might work for the secure store APIs (especially if we keep them internal). But at the same time, it does not feel very nice to me. The best would be to have a common EE sub-module where we could put stuff like this. Or maybe I should just create a special non-hot-restart package in the hot-restart module, to keep the damage under control.

mmedenjak replied:
Right, I would agree it would be best if we had a common EE sub-module but I don't think this is a big issue since the classes are private.

vojtechtoman commented:
Added the `keySize` property to `EncryptionAtRestConfig` (used for encryption key generation)
Also clarified that encryption key size is in bits.
Fixed the Javadoc of SecureStore.

mdogan commented on hazelcast/src/main/java/com/hazelcast/config/AbstractBasicSymmetricEncryptionConfig.java:
```diff
@@ -0,0 +1,96 @@
+ *
+ * @param <T> the type of the configuration class
+ */
+public abstract class AbstractBasicSymmetricEncryptionConfig<T extends AbstractBasicSymmetricEncryptionConfig> {
```
Minor: Can we call just `AbstractSymmetricEncryptionConfig`?

vojtechtoman replied:
Will do.

mdogan commented on hazelcast/src/main/java/com/hazelcast/config/SecureStoreConfig.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Abstract Secure Store configuration class.
+ */
+public abstract class SecureStoreConfig {
```
Can you explain a bit more what's the purpose of _Secure Store_?

vojtechtoman replied:
Will do.

mdogan commented on hazelcast/src/main/java/com/hazelcast/spi/impl/securestore/package-info.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Contains classes/interfaces for the Hazelcast Secure Store SPI.
+ */
+package com.hazelcast.spi.impl.securestore;
```
Is this package needed in OSS side?

vojtechtoman replied:
I had it in the `hot-restart` sub-module first, but after a discussion with @mmedenjak, we decided we wanted to make the concept independent of Hot Restart to allow possible future reuse. Since the APIs must be available to the `hot-restart` sub-module, the only suitable place is in OS.
But having said that, I have no problem with moving the interfaces back under `hot-restart` since they are not used anywhere else at the moment (and can be moved out if really needed).

mdogan replied:
If you already discussed this, I am fine with that. 👍

vojtechtoman commented:
run-lab-run

mmedenjak commented on hazelcast-spring/src/main/resources/hazelcast-spring-4.0.xsd:
```diff
@@ -3899,6 +3899,181 @@
 
+    <xs:element name="secure-store-substitution-group" abstract="true"/>
+
+    <xs:complexType name="keystore-entry">
```
Is this `keystore-entry` used anywhere?

vojtechtoman replied:
Good catch! A leftover from the previous quarter, will remove it.

vojtechtoman replied:
Removed

mmedenjak commented on hazelcast/src/main/java/com/hazelcast/config/AbstractSymmetricEncryptionConfig.java:
```diff
@@ -0,0 +1,96 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
```
Minor - I was planning, as part of next package cleanup, to move all `Abstract*Config` classes to internal as I don't see the user needing to do anything with those. WDYT, should we move this one as well as `SecureStoreConfig`?

vojtechtoman replied:
Hm, is it really a good idea to let public classes/interfaces implement/extend internal classes/interfaces? At the very least, internal APIs are not included in the generated Javadocs (not sure how clever Javadoc is when it comes to 'inheriting' excluded documentation).

mmedenjak replied:
Talked with Marko, we can take the middle-ground: the abstract classes should be package-private to indicate they are not for public use.

mmedenjak commented:
Merged with unrelated test failure - https://github.com/hazelcast/hazelcast/issues/15677

vojtechtoman commented:
Thanks for the reviews, everyone!
